code/game/objects/structures/universal_generators/cellular_automata.dm

```diff
@@ -0,0 +1,214 @@
+
+here is an example of what it might look like:
+
+############################################################
+###....####################################.....############
+##......######################..#########.........##########
+##......#####################....#######...........####.####
+##......###################.........................##...###
+##......##################..........................###...##
+#........##############.............................###...##
+#........#############...............................#....##
+##.......##############..................................###
+##.......###..############..............................####
+##.......##....############.............................####
+#..............############...###........................###
+#...............###########..#####...............##.......##
+#................#################...............##.......##
+##.....#####..........###########....#..........###.......##
+##....#######...........########....###.........####......##
+##....#######............######....####........#####......##
+##....#######.............####....#####.......#####......###
+#......######..............###....####........####......####
+#.......######.............###...####.........###.......####
+#........#####.............###..####.....................###
+##........####..............#...####.....................###
+#####......##...................####.....................###
+######...........................##.....................####
+######..................................................####
+######.........###.....................####.............####
+######......#########.................######............####
+#######....#############.......##############.....###..#####
+##############################################..############
+############################################################
+*/
+
+#define CA_PERMAWALL 2
+#define CA_WALL 1
+#define CA_FLOOR 0
+
+/obj/structure/radial_gen/cellular_automata
+	name = "cellular automata"
+	var/turf/ca_wall
+	var/turf/ca_floor
+	var/mapgrid_width = 40
+	var/mapgrid_height = 21
+	var/percent_area_walls = 45
+	var/iterations = 5
+	var/watch
+	var/mapgrid_scale = 1
+	var/list/list_of_turfs = list()
+	var/list/mapgrid
+	var/could_not_place = 0
+
+/obj/structure/radial_gen/cellular_automata/Destroy()
+	..()
+	mapgrid.Cut()
+	list_of_turfs.Cut()
+
+/obj/structure/radial_gen/cellular_automata/deploy_generator(var/turf/bottomleft)
+	init_mapgrid() // first generates a random map
+	for(var/i = 1 to iterations)
+		make_caverns(i)
+	could_not_place = check_mapgrid(bottomleft)
+	if(!could_not_place)
+		apply_mapgrid_to_turfs(bottomleft)
+
+/obj/structure/radial_gen/cellular_automata/proc/init_mapgrid()
+	if(!mapgrid)
+		mapgrid = new/list(mapgrid_width,mapgrid_height)
+	else if(could_not_place)
+		could_not_place = 0
+		return
+	for(var/row = 1 to mapgrid_height)
+		for(var/column = 1 to mapgrid_width)
+			if(column == 1)
```

Comment:
Why do this inside of the for loop rather than before/after, and have the loop go from `2` to `(n - 1)`

Reply:
🤷‍♂️